site/src/lib/model/apiKey.model.ts

```diff
@@ -0,0 +1,222 @@
+import crypto from "crypto";
+import { Db, DBRef } from "mongodb";
+import { v4 as uuid } from "uuid";
+import { mustGetEnvVar } from "../../util/api";
+
+import { User } from "./user.model";
+
+type ApiKeyProperties = {
+  createdAt: Date;
+  displayName: string;
+  hashedString: string;
+  lastUsedAt?: Date | null;
+  publicId: string;
+  revokedAt?: Date | null;
+  salt: string;
+  useCount: number;
+  user: DBRef;
+};
+
+export type UserFacingApiKeyProperties = Omit<
+  ApiKeyProperties,
+  "hashedString" | "salt" | "user"
+>;
+
+type ApiKeyDocument = ApiKeyProperties;
+
+export class ApiKey {
+  createdAt: Date;
+  displayName: string;
+  hashedString: string;
+  lastUsedAt?: Date | null;
+  publicId: string;
+  revokedAt?: Date | null;
+  salt: string;
+  useCount: number;
+  user: DBRef;
+
+  static readonly COLLECTION_NAME = "bp-api-keys";
+
+  private static readonly HASHING_SECRET = mustGetEnvVar("HASHING_SECRET");
+
+  /**
+   * This prefix helps to identify keys for both humans and automated code-scanners.
+   * It aims to balance human-readability with avoiding false positives (which plain "blocks" might do).
+   */
+  private static readonly PREFIX = "b10ck5";
+
+  private static readonly KEY_FORMAT_REGEXP = new RegExp(
+    // e.g. b10ck5.e84a616cedc8822b02ac96761b068f19.91ea0c90-acee-4215-a01e-85a4438af0ea
+    `^${ApiKey.PREFIX}\\.[a-z0-9]{32}\\.[a-z0-9]{8}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{4}-[a-z0-9]{12}$`,
+  );
+
+  private constructor({
+    createdAt,
+    displayName,
+    hashedString,
+    lastUsedAt,
+    publicId,
+    revokedAt,
+    salt,
+    useCount,
+    user,
+  }: ApiKeyProperties) {
+    this.createdAt = createdAt;
+    this.displayName = displayName;
+    this.hashedString = hashedString;
+    this.lastUsedAt = lastUsedAt;
+    this.publicId = publicId;
+    this.revokedAt = revokedAt;
+    this.salt = salt;
+    this.useCount = useCount;
+    this.user = user;
+  }
+
+  private static hash({
+    publicId,
+    privateId,
+    salt,
+  }: {
+    publicId: string;
+    privateId: string;
+    salt: string;
+  }) {
+    const stringToHash = `${publicId}.${privateId}.${salt}`;
+
+    return crypto
+      .createHmac("sha256", this.HASHING_SECRET)
+      .update(stringToHash)
+      .digest("hex");
+  }
+
+  static async create(
+    db: Db,
+    params: { displayName: string; user: User },
+  ): Promise<string> {
+    const { displayName, user } = params;
+
+    const privateId = uuid();
+    const publicId = crypto.randomBytes(16).toString("hex");
```

Comment:
Non blocking - continuation of above comment, the `16` byte count here could be the same const divided by 2. Also perhaps making it clear that 1 byte = 2 hex digits so no one will change it by accident